File: products/pushci/internal/ai/claw_client.go
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

const clawEndpoint = "https://claw-gateway.broad-dew-49ad.workers.dev"

// ClawClient calls the Claw Gateway for AI-powered analysis.
type ClawClient struct {
	APIKey    string
	ProjectID string
	Endpoint  string
}

// NewClawClient creates a Claw Gateway client from env vars.
func NewClawClient(apiKey, projectID string) *ClawClient {
	endpoint := os.Getenv("CLAW_GATEWAY_URL")
	if endpoint == "" {
		endpoint = clawEndpoint
	}
	return &ClawClient{
		APIKey:    apiKey,
		ProjectID: projectID,
		Endpoint:  endpoint,
	}
}

// NewClawClientFromEnv creates a client reading all config from env.
func NewClawClientFromEnv() *ClawClient {
	return NewClawClient(
		os.Getenv("CLAW_API_KEY"),
		os.Getenv("CLAW_PROJECT_ID"),
	)
}

// IsConfigured returns true if API key and project ID are set.
func (c *ClawClient) IsConfigured() bool {
	return c.APIKey != "" && c.ProjectID != ""
}

type clawRequest struct {
	ProjectID string `json:"project_id"`
	System    string `json:"system,omitempty"`
	Prompt    string `json:"prompt"`
	MaxTokens int    `json:"max_tokens"`
}

type clawResponse struct {
	Result string `json:"result"`
	Error  string `json:"error,omitempty"`
}

// Prompt sends a system+user message through the Claw Gateway.
func (c *ClawClient) Prompt(ctx context.Context, system, user string, maxTokens int) (string, error) {
	if !c.IsConfigured() {
		return "", fmt.Errorf("claw gateway: CLAW_API_KEY or CLAW_PROJECT_ID not set")
	}
	if maxTokens <= 0 {
		maxTokens = 2000
	}
	body, _ := json.Marshal(clawRequest{
		ProjectID: c.ProjectID,
		System:    system,
		Prompt:    user,
		MaxTokens: maxTokens,
	})
	req, err := http.NewRequestWithContext(ctx, "POST", c.Endpoint+"/v1/prompt", bytes.NewReader(body))
	if err != nil {
		return "", fmt.Errorf("claw gateway: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+c.APIKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("claw gateway: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		b, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("claw gateway %d: %s", resp.StatusCode, b)
	}
	var result clawResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("claw gateway decode: %w", err)
	}
	if result.Error != "" {
		return "", fmt.Errorf("claw gateway: %s", result.Error)
	}
	if result.Result == "" {
		return "", fmt.Errorf("claw gateway: empty response")
	}
	return result.Result, nil
}

File: products/tenantiq/apps/api/src/lib/claw-client.ts
/**
 * Claw Gateway Client — routes AI requests through the shared gateway
 * for centralized billing, usage tracking, rate limiting, and prompt guard.
 *
 * Endpoint: https://claw-gateway.broad-dew-49ad.workers.dev
 * Falls back to direct Anthropic API if gateway is unreachable.
 */

import { AI } from './constants';

const CLAW_GATEWAY_URL = 'https://claw-gateway.broad-dew-49ad.workers.dev';
const TIMEOUT_MS = 30_000;

export interface ClawConfig {
	apiKey: string;
	projectId: string;
	gatewayUrl?: string;
}

export interface ClawPromptRequest {
	prompt: string;
	system?: string;
	provider?: 'anthropic' | 'openai' | 'groq';
	model?: string;
	maxTokens?: number;
	temperature?: number;
	metadata?: Record<string, string>;
}

export interface ClawPromptResponse {
	text: string;
	model: string;
	usage: { inputTokens: number; outputTokens: number };
	cached: boolean;
	durationMs: number;
}

export class ClawClient {
	private apiKey: string;
	private projectId: string;
	private gatewayUrl: string;

	constructor(config: ClawConfig) {
		this.apiKey = config.apiKey;
		this.projectId = config.projectId;
		this.gatewayUrl = config.gatewayUrl || CLAW_GATEWAY_URL;
	}

	/** Send a prompt through the Claw gateway. */
	async prompt(req: ClawPromptRequest): Promise<ClawPromptResponse> {
		const controller = new AbortController();
		const timeout = setTimeout(() => controller.abort(), TIMEOUT_MS);

		try {
			const provider = req.provider || 'anthropic';
			// Opus 4.7 (and any Claude 4.x model the gateway forwards to) rejects
			// temperature/top_p/top_k. Only include temperature when the caller
			// asked for it AND the provider is non-Anthropic.
			const payload: Record<string, unknown> = {
				prompt: req.prompt,
				system: req.system,
				provider,
				model: req.model || AI.MODEL,
				maxTokens: req.maxTokens || AI.MAX_TOKENS_DEFAULT,
				metadata: { ...req.metadata, source: 'tenantiq' },
			};
			if (req.temperature !== undefined && provider !== 'anthropic') {
				payload.temperature = req.temperature;
			}
			const res = await fetch(`${this.gatewayUrl}/v1/prompt`, {
				method: 'POST',
				headers: {
					'Content-Type': 'application/json',
					Authorization: `Bearer ${this.apiKey}`,
					'X-Project-Id': this.projectId,
				},
				body: JSON.stringify(payload),
				signal: controller.signal,
			});

			if (!res.ok) {
				const body = await res.text().catch(() => '');
				throw new Error(`Claw gateway ${res.status}: ${body}`);
			}

			return (await res.json()) as ClawPromptResponse;
		} finally {
			clearTimeout(timeout);
		}
	}

	/** Health check — verify the gateway is reachable. */
	async health(): Promise<{ ok: boolean; latencyMs: number }> {
		const start = Date.now();
		try {
			const res = await fetch(`${this.gatewayUrl}/health`, {
				headers: { Authorization: `Bearer ${this.apiKey}` },
			});
			return { ok: res.ok, latencyMs: Date.now() - start };
		} catch {
			return { ok: false, latencyMs: Date.now() - start };
		}
	}
}

/**
 * Create a ClawClient from environment variables.
 * Returns null if CLAW_API_KEY is not set.
 */
export function createClawClient(env: {
	CLAW_API_KEY?: string;
	CLAW_PROJECT_ID?: string;
	CLAW_GATEWAY_URL?: string;
}): ClawClient | null {
	if (!env.CLAW_API_KEY) return null;
	return new ClawClient({
		apiKey: env.CLAW_API_KEY,
		projectId: env.CLAW_PROJECT_ID || 'tenantiq',
		gatewayUrl: env.CLAW_GATEWAY_URL,
	});
}
